src/null-safety/migrating-to-null-safety/index.md

```diff
@@ -0,0 +1,435 @@
+every SDK will continue to support legacy code that does *not* use null safety.
+
+Since the language version is derived from the SDK version, it implies:
+
+*   Whenever we ship a minor version of the SDK, a new language version appears.
+    In practice, many of these language versions are very similar to and
+    entirely compatible with previous versions. For example, Dart 2.9 (the
+    language) is essentially identical to 2.8.
+
+*   When we ship a patch version of the SDK, it cannot contain any backwards
+    incompatible language changes. Since a patch SDK release (like 2.7.2) does
+    not introduce a new language version (2.7) over the previous SDK (2.7.1,
+    also language 2.7), there would be no way to avoid breaking user code if
+    we did.
+
+### Specifying language version
+
+Every Dart SDK from **TODO: version** and later supports null safety. But they
+also all still support older null unsafe "legacy" code. To keep that code
+running, the SDK just needs to know that the code is targeting an older Dart
+language version. How do you tell the compiler that?
+
+We wanted to make the process of pinning your code to a language version as
+lightweight as possible. We wouldn't want to, say, use a different file
+extension like `.dartx` and make you rename every single one of your files every
+time a new language version came out. Instead, we build on top of what the tools
+already know about the package containing your code.
+
+Your [pubspec] probably already has an [SDK constraint] like:
+
+[pubspec]: https://dart.dev/tools/pub/pubspec
+[sdk constraint]: https://dart.dev/tools/pub/pubspec#sdk-constraints
+
+```dart
+environment:
+  sdk: '>=2.6.0 <3.0.0'
+```
+
+Here, your package claims that it can be run using any Dart SDK from 2.5.0 or
+later, until 3.0.0. For that to be true, your package can't be using any
+features that were introduced after 2.6.0. If you used, say, extension methods
+(released in 2.7.0), then your package wouldn't work if someone tried to run it
+on a Dart 2.6.0 SDK, which your own pubspec claims is supported.
+
+Our tools are smart enough to understand what that SDK constraint implies and
+use that to infer a language version for the libraries in your package. The rule
+is simple: **The default Dart language version for a package is the lowest
+language version covered by its SDK constraint.** So, in this case, Dart tools
+assume the code in your package is targeting Dart 2.6 and won't let you use any
+features not supported by that version.
+
+Almost every existing pubspec out there has an SDK constraint whose range
+includes Dart SDKs prior to null safety. Using those lower bounds to select the
+language version automatically opts all of them out of null safety, which keeps
+them running just as they did before.
+
+### Language comment markers
+
+Note that I said the *default* language version for code in your package. Tools
+use your pubspec (or more precisely, the package config file generated from your
```

Comment:
Can we link 'package config file' to something?

Maybe https://pub.dev/packages/package_config ?

Reply:
Done.